modules/xphoto/src/oilpainting.cpp

```diff
@@ -0,0 +1,129 @@
+                            if (y + yy >= 0 && y + yy < imgSrc.rows)
+                            {
+                                Vec3b *vPtr = (Vec3b *)imgSrc.ptr(y + yy) + x - 0;
+                                uchar *uc = imgLuminance.ptr(y + yy) + x - 0;
+                                for (int xx = 0; xx <= halfsize; xx++, vPtr++, uc++)
+                                {
+                                    if (x + xx >= 0 && x + xx < imgSrc.cols)
+                                    {
+                                        histogram[*uc]++;
+                                        meanBGR[*uc] += *vPtr;
+                                    }
+                                }
+                            }
+                        }
+
+                    }
+                    else
+                    {
+                        for (int yy = -halfsize; yy <= halfsize; yy++)
+                        {
+                            if (y + yy >= 0 && y + yy < imgSrc.rows)
+                            {
+                                Vec3b *vPtr = (Vec3b *)imgSrc.ptr(y + yy) + x - halfsize - 1;
+                                uchar *uc = imgLuminance.ptr(y + yy) + x - halfsize - 1;
+                                int xx = -halfsize - 1;
+                                if (x + xx >= 0 && x + xx < imgSrc.cols)
+                                {
+                                    histogram[*uc]--;
+                                    meanBGR[*uc] -= *vPtr;
+                                }
+                                vPtr = (Vec3b *)imgSrc.ptr(y + yy) + x + halfsize;
+                                uc = imgLuminance.ptr(y + yy) + x + halfsize;
+                                xx = halfsize;
+                                if (x + xx >= 0 && x + xx < imgSrc.cols)
+                                {
+                                    histogram[*uc]++;
+                                    meanBGR[*uc] += *vPtr;
+                                }
+                            }
+                        }
+                    }
+                    auto pos = distance(histogram.begin(), std::max_element(histogram.begin(), histogram.end()));
+                    *vDst = meanBGR[pos] / histogram[pos];
+                }
+            }
+        }
+    };
+
+    class Quotient {
+        uchar q;
+    public:
+        Quotient(int v) { q = static_cast<uchar>(v); };
+        void operator ()(uchar &pixel, const int * ) const {
+            pixel = pixel / q;
+        }
+    };
+
+    void oilPainting(InputArray _src, OutputArray _dst, int size, int dynValue)
+    {
+        oilPainting(_src, _dst, size, dynValue, COLOR_BGR2GRAY);
+    }
+
+    void oilPainting(InputArray _src, OutputArray _dst, int size, int dynValue,int code)
+    {
+        CV_Assert(_src.kind() == _InputArray::MAT && size>=1 && dynValue>0 && dynValue<128);
+        Mat src = _src.getMat();
+        CV_Assert(src.type()==CV_8UC1 || src.type() == CV_8UC3);
```

Comment:
Please use `CV_CheckTypeEQ(src.type(), src.type()==CV_8UC1 || src.type() == CV_8UC3, "")` to provide useful messages for OpenCV users.

Assertion above should be spitted too. "int" values should be validated by `CV_Check*()` macros.